src/string_util.cpp

```diff
@@ -400,12 +401,21 @@ namespace libtorrent {
 
 	bool is_i2p_url(std::string const& url)
 	{
+		static std::unordered_map<std::string, bool> cache;
```

Comment:
I'm not sure my concern is valid but anyway:
1. What do you think about putting an upper limit on how many entries in the cache? IMO it isn't ideal for it to be unlimited. I suppose `1024` would suffice for most cases. Also this implies it needs some kind of cache replacement policy, for example LRU. Maybe libtorrent/boost provides such container?
2. If I understand correctly, using `static` means it will be shared across all libtorrent sessions which isn't ideal either. It should be local to a specific torrent/session or whatever scope that make sense.
3. Maybe this function requires to be thread-safe.

Reply:
1. It looks like libtorrent implements an LRU cache in [src/file_view_pool.cpp](https://github.com/arvidn/libtorrent/blob/RC_2_0/src/file_view_pool.cpp), but that's specific to the file pool. [Boost also has an implementation](https://www.boost.org/doc/libs/1_67_0/boost/compute/detail/lru_cache.hpp), but it's under ::compute, so it can't be used here. I could implement one, but I'm not sure if it's needed - my qbittorrent uses 242 trackers urls in total, being ~8kb in text. Round it up to 10kb to account for `std::unordered_map`'s overhead, and it shouldn't get that big.

2 & 3. I hadn't considered multiple threads, but I think I could just use `thread_local` instead of `static`.

Reply:
I think this cache is the only controversial part of this patch. It seems more natural to record whether a tracker is i2p or not once, and not call this function repeatedly. It's only 1 bit of information.
The ABI issue is a challenge though. in `master` the announce entries have been separated into internal and exported types, so it's easy to add internal fields without affecting the ABI

Reply:
Since it's only one bit, could it be put in padding, [like `verified`](https://github.com/arvidn/libtorrent/blob/9eaa8522cb6da2cb2ae8277d95b183bcfc5adbc9/include/libtorrent/aux_/announce_entry.hpp#L192)? Or could that cause some issues?